### ISSUE TITLE
Proxy-vole crashes in JNA/native code on Windows 11 and Java 8

### ISSUE THREAD

**Issue by dtuma** [akuhtz/proxy-vole#37]
Using Java 8 u331 on Windows 11 with proxy-vole 1.0.16, I'm seeing crashes in native code nearly 100% of the time. From the crash dump:

```
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.sun.jna.Native._getPointer(J)J+0
j  com.sun.jna.Native.getPointer(J)Lcom/sun/jna/Pointer;+1
j  com.sun.jna.Pointer.getPointer(J)Lcom/sun/jna/Pointer;+6
j  com.sun.jna.Memory.getPointer(J)Lcom/sun/jna/Pointer;+11
j  com.github.markusbernhardt.proxy.jna.win.WTypes2$LPWSTRByReference.getPointerToString()Lcom/sun/jna/Pointer;+5
j  com.github.markusbernhardt.proxy.jna.win.WTypes2$LPWSTRByReference.finalize()V+1
J 3298 C2 java.lang.ref.Finalizer.access$100(Ljava/lang/ref/Finalizer;Lsun/misc/JavaLangAccess;)V (6 bytes) @ 0x0000024580222d5c [0x0000024580222b80+0x1dc]
j  java.lang.ref.Finalizer$FinalizerThread.run()V+45
v  ~StubRoutines::call_stub
```

The problem seems to go away if I use a newer version of Java. But Java 8 is commonly deployed by my userbase due to the changes in Oracle licensing for Java 9 and up.

Any insights into what might be going on?

**Comment by akuhtz** [comment]
Hello David,
Thanks for reporting. There might be an issue WTypes2$LPWSTRByReference where the finalize method is implemented.

**Comment by akuhtz** [comment]
@dtuma Can you try with the current ```1.0.17-SNAPSHOT``` version if the issue still exists?

**Comment by dtuma** [comment]
I checked out the latest version of master (which upgrades JNA to 5.11.0). There was no change from the previous behavior.

I also tried the fix-37 branch; but the unit tests crash in native code, causing the mvn build to fail. The crashing test class was `com.github.markusbernhardt.proxy.selector.pac.PacScriptMethodsTest` but I'm having trouble finding any dump file.

One possibly relevant detail: these crashes are occurring on a computer that does not use any proxy at all (direct connection to the internet). If you  make a call to WinHttp to read a proxy-related string, and no proxy is in effect, is it a mistake to free the memory associated with the pointer?

**Comment by akuhtz** [comment]
> I also tried the fix-37 branch; but the unit tests crash in native code, causing the mvn build to fail. The crashing test class was `com.github.markusbernhardt.proxy.selector.pac.PacScriptMethodsTest` but I'm having trouble finding any dump file.
> 
Have you checked the ```target/surefire-reports``` directory?

> One possibly relevant detail: these crashes are occurring on a computer that does not use any proxy at all (direct connection to the internet). If you make a call to WinHttp to read a proxy-related string, and no proxy is in effect, is it a mistake to free the memory associated with the pointer?

I pushed a change in branch ```fix-37``` where the memory is only freed if the return value of the call is true.

As I cannot reproduce it here it might be useful to set the Slf4jLogBackEnd in the test to better see what's going on:
```com.github.markusbernhardt.proxy.util.Logger.setBackend(new Slf4jLogBackEnd());```

**Comment by akuhtz** [comment]
I pushed a change to branch ```fix-37``` that reverts a mistake I did. Can you try again?

**Comment by dtuma** [comment]
Thanks for your help. I've pulled in your latest change, and I've done a bunch of testing on my own as well. I think I have some answers about the root cause and the current fix. It's complicated, so I hope I can explain it well.

The native errors were occurring during finalization, so they happen when the garbage collector decides a `WTypes2.LPWSTRByReference` object is no longer reachable. Finalization happens at unspecified times and is inherently JVM-specific. That makes these errors unpredictable and tricky to debug.

But `WTypes2.LPWSTRByReference` is not the only class with a finalizer. The `com.sun.jna.Native` class has a static finalizer that runs when the class definition is GC'd. That finalizer cleans up all of the native resources created by JNA, including every instance of `com.sun.jna.Memory` that was created by the JVM.

One interesting detail about finalization is that when an entire tree of objects becomes unreachable, the garbage collector can finalize all the objects in that that tree concurrently, and/or in any order. So when I saw errors, it was because the JVM decided to run the `com.sun.jna.Native` finalizer first, followed by the `WTypes2.LPWSTRByReference` finalizer. 

When the `com.sun.jna.Native` finalizer runs, it aggressively frees the memory associated with every `com.sun.jna.Memory` object. This includes a `Memory` object that's a field of LPWSTRByReference. When the `LPWSTRByReference` finalizer ran later, it was attempting to access that already-disposed `Memory` object, triggering a segfault.

Your current fix - freeing the memory explicitly after making the `WinHttpDetectAutoProxyConfigUrl` call - is much better. It makes a native call and then immediately cleans up. That's much better than waiting until finalization, when some of the objects needed might have been cleaned up before us.

I created a test harness that could reliably trigger the native crash, and your current fix appears to fix the problem. (BUT my code is not exercising the new `Kernel32Util.freeGlobalMemory` instruction because I'm not behind a WPAD-configured proxy; so I can't vouch for the correctness of that line.)

I'm still seeing some very sporadic native crashes that are unrelated to the `WTypes2.LPWSTRByReference` finalizer. I'm still trying to track those down, but I wanted to give you an update on what I've learned so far.

**Comment by akuhtz** [comment]
Thanks for testing and providing feedback. I have setup a VM where I can change the proxy settings and I can see my (debug) error message when testing with the ```java -jar proxy-vole-1.0.17-SNAPSHOT-jar-with-dependencies.jar``` (that invokes ```ProxyTester```) in the target directory.

The finalizer are deprecated in Java 9 and I think it's much better to clean up the memory directly and not leave it up to the finalizer that might even not runs unpredictable or event don't run on some implementations. I looked at the sources of jna itself (e.g. ```Advapi32Util```) and they free local memory immediately after using it, too (it might have changed to this over time).

There are more calls that need to free memory according to Windows documentation (like https://docs.microsoft.com/en-us/windows/win32/api/winhttp/nf-winhttp-winhttpgetdefaultproxyconfiguration) but I have not touched it yet.

**Comment by dtuma** [comment]
> There are more calls that need to free memory according to Windows documentation but I have not touched it yet.

I noticed that too. I had started to code those changes, I'd be happy to offer a pull request if you like.

I also noticed, that since you removed the finalizer from `LPWSTRByReference`, that class no longer provides much to distinguish itself from `com.sun.jna.ptr.PointerByReference`. The logic it does provide has a number of indirections that make it tricky to follow. Would you be open to a pull request that removes the class and uses `PointerByReference` instead?

**Comment by akuhtz** [comment]
Sure, PR are welcome.

**Comment by akuhtz** [comment]
@dtuma  Released with 1.0.17. Thanks for the PR and testing!
